### ISSUE TITLE
Activitypub broadcast jobs pile up

### ISSUE THREAD

**Issue by scanlime** [Chocobozzz/PeerTube#4249]
This is essentially a repeat of bug #4184 but for broadcast jobs rather than fetch jobs:

![image](https://user-images.githubusercontent.com/923812/125341167-2b433380-e308-11eb-9efa-dead5c7be787.png)

Would it be safe to bump the broadcast job concurrency as well?

**Comment by scanlime** [comment]
From my limited understanding of ActivityPub and the http broadcaster, it seems like most of the jobs are independent and could be parallelized, but it is possible for there to be ordering dependencies. For example, if one video's description is updated multiple times, my understanding is that it's important for those posts to go out in order.

I'm trying to look for additional quick solutions, especially something I can do without restarting the server. Are other larger servers hitting this issue too? I would expect it to be a common scalability issue, given how chatty the protocol is. These messages are generated for any update but also for video views, and they go out to every follower.

**Comment by scanlime** [comment]
For some more context, here's the last 30 days of diode.zone's activitypub-http-broadcast queue depth. It looks like we are usually clearing the queue just a little faster than we're filling it, but that balance is thrown off occasionally and we end up with these deep queues, resulting in very high latency of many hours for publishing new videos to the fediverse.

![image](https://user-images.githubusercontent.com/923812/125350232-58e1aa00-e313-11eb-9f1b-5c471acf95fe.png)

**Comment by scanlime** [comment]
As a note for other sysadmins hitting this: Removing a slow follower is actually not a sufficient solution to this, since the follower list is included in the queued job. All of the jobs already backed up in the queue will be trying to contact whatever stuck server is causing broadcasts to take their maximum length. (On my system, an unreachable server times out after 2 minutes)

I worked around the clog by adding a firewall rule to quickly reject connection attempts to the misbehaving peer. The score system in peertube should eventually react to situations like this, but it clearly doesn't react fast enough when the broadcast queue introduces latency between the decision-making and the actual broadcast.

So, I would watch the netstat output to look for the connections that stay stuck during the entire broadcast cycle, and then I added those IPs temporarily to a set of misbehaving peers in my firewall config.

**Comment by scanlime** [comment]
I'm not sure if this warrants a separate issue or generalizing this one, but activitypub refresher jobs are piling up now. This seems to be the result of some downtime at peervideo.club.

![image](https://user-images.githubusercontent.com/923812/125838007-866aad39-5425-4560-9b84-0b779c9327fe.png)

This still feels kludgy to me, but in the hopes of a slightly more general solution to keep activitypub from slowing down when one peer goes down I'm adjusting the kernel's SYN retries (down from 6 to 4):

```
sysctl net.ipv4.tcp_syn_retries=4
```

Taking a higher level view though, is it really intentional that we're trying to deliver messages in lockstep amongst all peers, allowing one slow or broken peer to gunk up the entire sync process?

Would it be a reasonable architectural goal to restructure the concurrency of activitypub sync'ing so that each peer can go at its own pace?

How much does PeerTube care about reliably delivering all updates to all peers?

**Comment by Chocobozzz** [comment]
Hello,

Thanks for the details.

Could you try to upgrade to 3.3-rc.1 (or wait the stable next week)? Or apply [this patch](https://github.com/Chocobozzz/PeerTube/commit/bbb3be686ad960c5b1f34a7771dec09eb1783455) to a 3.2 running instance. This fixes a weird bugs and should improve federation.

If it does not fix this job pile, I think we can improve the broadcast delivery architecture.

**Comment by scanlime** [comment]
Hi Chocobozzz,

I am already running 3.3.0-rc1. I'm fairly sure this is not a DNS issue, as if i watch the output of netstat I can see the hung connections waiting on SYN, they've already been resolved to an IP. I've been trying to resolve this by changing timeouts or looking for individual broken servers, but it does look like the root cause here is just the lockstep delivery algorithm combined with the inherent unreliability of large numbers of computers.

**Comment by scanlime** [comment]
Related to the views update interval, the interaction between view updates and the activitypub broadcast has been causing a lot of backlog lately. You can see here there are only brief CPU load spikes corresponding to the view updates, but a notable backlog in broadcasts every hour in the Job Queue graph,

![image](https://user-images.githubusercontent.com/923812/134430912-be4af17d-ad59-46ce-9c18-2406b772d137.png)

**Comment by kontrollanten** [comment]
Thanks for your work with this, @scanlime Do I understand that this issue is due to slow followers that respond slowly to the broadcast messages? Could a solution be to create one job per follower? A long term solution could be to rank the followers by response time, and maybe send multiple jobs to a follower in one request (when there's a backlog)?

Do you know what's causing the CPU spikes during view updates?

**Comment by scanlime** [comment]
I've been on the lookout for slow followers and that's been the problem once or twice, but at this point it's just taking too long to deliver all the messages serially even when the followers are fairly well behaved.

I think we just need a different queueing strategy, where many followers can be processed in parallel (limited by desired RAM usage and concurrent tcp connections) yet still get their messages in order. Sending as much of that queue as possible in large batches would help too.

I'm not sure the CPU usage itself is a problem, but it might be helpful to reduce the spikiness of the view updates by batching them differently, not just unleashing them all on the top of the hour.
